lib/Value/Matrix.pm

```diff
@@ -749,27 +952,171 @@ sub column {
 	}
 	return if $j + 1 > $d[1] || $j < -$d[1];
 	my @col = ();
-	foreach my $row (@{ $M->data }) { push(@col, $self->make($row->data->[$j])) }
+	for my $row (@{ $M->data }) { push(@col, $self->make($row->data->[$j])) }
 	return $self->make(@col);
 }
 
-#
-#  Extract a given element from the matrix
-#
+=head3 C<element>
+
+Extract an element from the given row/col.
+
+Usage:
+
+    my $A    = Matrix([ [ 1, 2, 3, 4 ], [ 5, 6, 7, 8 ], [ 9, 10, 11, 12 ] ]);
+    $A->element(2,3); # returns 7
+
+    my $B = Matrix([ [ [ 1, 2 ], [ 3, 4 ] ], [ [ 5, 6 ], [ 7, 8 ] ] ]);
+    $B->element(1,2,1); # returns 3;
+
+    my $row = Matrix([4,3,2,1]);
+    $row->element(2); # returns 3;
+=cut
+
 sub element {
 	my $self = (ref($_[0]) ? $_[0] : shift);
 	my $M    = $self->promote(shift);
 	return $M->extract(@_);
 }
 
-# @@@ assign @@@
+=head3 C<setElement>
+
+Assign an element in the matrix to a value.
+
+Inputs: indices as an arrayref and the value in a form that can be parsed.
+
+Note: this mutates the matrix itself.
+
+Usage:
+
+    my $A = Matrix([ [ 1, 2, 3, 4 ], [ 5, 6, 7, 8 ], [ 9, 10, 11, 12 ] ]);
+    $A->setElement([2,3],-5);
+
+
+=cut
+
+sub setElement {
+	my ($self, $ind, $value) = @_;
+
+	# Drill down into the matrix
+	my $el = $self->{data}[ $ind->[0] - 1 ];
+	for my $i (1 .. scalar(@$ind) - 1) { $el = $el->{data}[ $ind->[$i] - 1 ]; }
+
+	# update the value of $el
+	$el = Value::makeValue($value);
+}
+
+=head3 C<subMatrix>
+
+
+Return a submatrix of the matrix.  If the rows and columns are array refs, the given rows and
+columns of the matrix are returns as a Matrix object.
+
+If the input are integers, then the submatrix with that row and column removed.
+
+Usage:
+
+    $A = Matrix([[1,2,3,4],[5,6,7,8],[9,10,11,12]]);
+
+    $A->subMatrix([2..3],[2..4]);  # returns a Matrix([[6,7,8],[10,11,12]])
+
+    $A->subMatrix(2,3);  # returns Matrix([ [ 1, 2, 4 ], [ 9, 10, 12 ] ]);
+
+    $A->subMatrix([3,1,2],[1,4,2]);  # returns Matrix([9,12,10],[1,4,2],[5,8,6]);
+=cut
+
+sub subMatrix {
```

Comment:
This has some issues as well. If I try it on a 1D matrix, I get an error:
```
$A = Matrix([ 1, 2, 3, 4 ]);
$B = $A->subMatrix([3]);
```

And if I try it on a 3D matrix, it is not returning what I would expect it to:
```
$A = Matrix([ [[1,2],[3,4]], [[5,6], [7,8]]]);
$B = $A->subMatrix([1], [1,2], [2]);
# $B seems to be what I would expect from subMatrix([1], [1,2], [1,2]);
```

Reply:
I thought of subMatrix only on 2D matrices, so will rework for other dimensions.

Reply:
Rewrote this now.  Should be working on non-2D matrices.